### PR TITLE
Traversals in folds for PostgreSQL

### PR DESCRIPTION
**Pull request by chewselene**
This PR implements traversals inside of folds for PostgreSQL. The `SQLFoldObject` methods `visit_traversed_vertex` and `visit_output_vertex` have been combined into a single method `visit_vertex`, which performs a traversal and checks if there are any outputs at the new location. Since outputs can only come from a single location within a fold, `visit_vertex` throws an error if the user attempts to output from multiple locations within the fold.

### CODE REVIEW

File: graphql_compiler/tests/test_compiler.py [hunk 1]
```diff
@@ -5872,15 +5852,35 @@ def test_fold_and_traverse(self):
...
+                    array_agg("Animal_3".name) AS fold_output_name
+                FROM schema_1."Animal" AS "Animal_2"
+                JOIN schema_1."Animal" AS "Animal_4"
+                ON "Animal_2".parent = "Animal_4".uuid
+                JOIN schema_1."Animal" AS "Animal_3"
+                ON "Animal_4".uuid = "Animal_3".parent
+                GROUP BY "Animal_2".uuid
+            ) AS folded_subquery_1
+            ON "Animal_1".uuid = folded_subquery_1.uuid
+        """
```
Comment (chewselene) [inline graphql_compiler/tests/test_compiler.py:5874]:
```
    {
        Animal {
            name @output(out_name: "animal_name")
            in_Animal_ParentOf @fold {
                out_Animal_ParentOf {
                    name @output(out_name: "sibling_and_self_names_list")
                }
            }
        }
    }
```

File: graphql_compiler/tests/test_compiler.py [hunk 2]
```diff
@@ -5949,15 +5949,36 @@ def test_fold_and_deep_traverse(self):
...
+                JOIN schema_1."Animal" AS "Animal_3"
+                ON "Animal_2".parent = "Animal_3".uuid
+                JOIN schema_1."Animal" AS "Animal_4"
+                ON "Animal_3".uuid = "Animal_4".parent
+                JOIN schema_1."Species" AS "Species_1"
+                ON "Animal_4".species = "Species_1".uuid
+                GROUP BY "Animal_2".uuid
+            ) AS folded_subquery_1
+            ON "Animal_1".uuid = folded_subquery_1.uuid
+            """
```
Comment (chewselene) [inline graphql_compiler/tests/test_compiler.py:5972]:
```
    {
        Animal {
            name @output(out_name: "animal_name")
            in_Animal_ParentOf @fold {
                out_Animal_ParentOf {
                    out_Animal_OfSpecies {
                        name @output(out_name: "sibling_and_self_species_list")
                    }
                }
            }
        }
    }
```

File: graphql_compiler/tests/test_compiler.py [hunk 3]
```diff
@@ -6232,15 +6274,37 @@ def test_multiple_outputs_in_same_fold_and_traverse(self):
...
+                    array_agg("Animal_3".name) AS fold_output_name
+                FROM schema_1."Animal" AS "Animal_2"
+                JOIN schema_1."Animal" AS "Animal_4"
+                ON "Animal_2".parent = "Animal_4".uuid
+                JOIN schema_1."Animal" AS "Animal_3"
+                ON "Animal_4".uuid = "Animal_3".parent
+                GROUP BY "Animal_2".uuid
+            ) AS folded_subquery_1
+            ON "Animal_1".uuid = folded_subquery_1.uuid
+        """
```
Comment (chewselene) [inline graphql_compiler/tests/test_compiler.py:6298]:
```
    {
        Animal {
            name @output(out_name: "animal_name")
            in_Animal_ParentOf @fold {
                out_Animal_ParentOf {
                    name @output(out_name: "sibling_and_self_names_list")
                    uuid @output(out_name: "sibling_and_self_uuids_list")
                }
            }
        }
    }
```

File: graphql_compiler/tests/test_compiler.py [hunk 4]
```diff
@@ -6546,15 +6610,54 @@ def test_multiple_folds_and_traverse(self):
...
+                    array_agg("Animal_6".name) AS fold_output_name
+                FROM schema_1."Animal" AS "Animal_5"
+                JOIN schema_1."Animal" AS "Animal_7"
+                ON "Animal_5".parent = "Animal_7".uuid
+                JOIN schema_1."Animal" AS "Animal_6"
+                ON "Animal_7".uuid = "Animal_6".parent
+                GROUP BY "Animal_5".uuid
+            ) AS folded_subquery_2
+            ON "Animal_1".uuid = folded_subquery_2.uuid
+        """
```
Comment (chewselene) [inline graphql_compiler/tests/test_compiler.py:6651]:
```
    {
        Animal {
            name @output(out_name: "animal_name")
            out_Animal_ParentOf @fold {
                in_Animal_ParentOf {
                    name @output(out_name: "spouse_and_self_names_list")
                    uuid @output(out_name: "spouse_and_self_uuids_list")
                }
            }
            in_Animal_ParentOf @fold {
                out_Animal_ParentOf {
                    name @output(out_name: "sibling_and_self_names_list")
                    uuid @output(out_name: "sibling_and_self_uuids_list")
                }
            }
        }
    }
```

File: graphql_compiler/tests/test_compiler.py [hunk 5]
```diff
@@ -9967,15 +10089,40 @@ def test_optional_traversal_and_fold_traversal(self):
...
+                FROM schema_1."Animal" AS "Animal_4"
+                JOIN schema_1."Animal" AS "Animal_6"
+                ON "Animal_4".uuid = "Animal_6".parent
+                JOIN schema_1."Animal" AS "Animal_5"
+                ON "Animal_6".uuid = "Animal_5".parent
+                GROUP BY "Animal_4".uuid
+            ) AS folded_subquery_1
+            ON "Animal_1".uuid = folded_subquery_1.uuid
+            WHERE "Animal_2".uuid IS NOT NULL OR "Animal_3".uuid IS NULL
+"""
```
Comment (chewselene) [inline graphql_compiler/tests/test_compiler.py:10116]:
```
    {
        Animal {
            name @output(out_name: "animal_name")
            in_Animal_ParentOf @optional {
                in_Animal_ParentOf {
                    name @output(out_name: "grandparent_name")
                }
            }
            out_Animal_ParentOf @fold {
                out_Animal_ParentOf {
                    name @output(out_name: "grandchild_names_list")
                }
            }
        }
    }
```

File: graphql_compiler/tests/test_compiler.py [hunk 6]
```diff
@@ -10079,15 +10226,40 @@ def test_fold_traversal_and_optional_traversal(self):
...
+                ON "Animal_5".uuid = "Animal_4".parent
+                GROUP BY "Animal_3".uuid
+            ) AS folded_subquery_1
+            ON "Animal_1".uuid = folded_subquery_1.uuid
+            LEFT OUTER JOIN schema_1."Animal" AS "Animal_6"
+            ON "Animal_1".parent = "Animal_6".uuid
+            LEFT OUTER JOIN schema_1."Animal" AS "Animal_2"
+            ON "Animal_6".parent = "Animal_2".uuid
+            WHERE "Animal_2".uuid IS NOT NULL OR "Animal_6".uuid IS NULL
+            """
```
Comment (chewselene) [inline graphql_compiler/tests/test_compiler.py:10253]:
```
    {
        Animal {
            name @output(out_name: "animal_name")
            out_Animal_ParentOf @fold {
                out_Animal_ParentOf {
                    name @output(out_name: "grandchild_names_list")
                }
            }
            in_Animal_ParentOf @optional {
                in_Animal_ParentOf {
                    name @output(out_name: "grandparent_name")
                }
            }
        }
    }
```

File: graphql_compiler/tests/test_compiler.py [hunk 7]
```diff
@@ -5834,27 +5834,7 @@ def test_fold_and_traverse(self):
...
-        #         FROM db_1.schema_1.[Animal] AS [Animal_2]
-        #         JOIN db_1.schema_1.[Animal] AS [Animal_3]
-        #         ON [Animal_2].parent = [Animal_3].uuid
-        #         JOIN db_1.schema_1.[Animal] AS [Animal_4]
-        #         ON [Animal_3].uuid = [Animal_4].parent
-        #         GROUP BY [Animal_2].uuid
-        #     ) AS folded_subquery_1
-        #     ON [Animal_1].uuid = folded_subquery_1.uuid
-        # '''
+        expected_mssql = NotImplementedError
```
Comment (obi1kenobi) [inline graphql_compiler/tests/test_compiler.py:5837]: This one isn't supported at the moment?

Comment (chewselene) [reply graphql_compiler/tests/test_compiler.py:5837]: This commented out `expected_sql` that got deleted is largely similar to the `expected_postgresql` below. Traversals inside of folds are not yet implemented for MSSQL, but the `TODO` needs to be updated to reflected that.

**Comment by chewselene** [comment]
> * Is there a potential for name conflicts between user-specified output names and the output we use as the join key between the fold's enclosing scope and its first scope within the fold?

discussed offline, but the tldr is that the `fold_output_{field_name}` intermediate output names are generated so the user cannot cause a duplicate name. The user-specified output name will be used to name the `COALESCE` output, but the user has no control over output names in the folded subquery.

> * The vast majority of this part of the codebase is still untyped, and it would be nice to start transitioning it to typed. A heuristic I like is that if you make changes to a function, you hopefully know its inputs/outputs, so make sure to add them to the type signature. If you feel that doing so will bloat this diff's line count, I'm happy to wait for a separate PR, but it would be nice to do sooner rather than later.

+1000 to this idea. Will keep it to a separate PR because I think it'll make it more difficult to review here, but will definitely be creating a PR to update type hinting and documentation with no functional changes next.

File: graphql_compiler/compiler/emit_sql.py [hunk 8]
```diff
@@ -693,16 +706,34 @@ def __init__(self, sql_schema_info, ir):
...
-        """Move to a different location in the query, updating the _alias."""
+        """Move to a different location in the query, updating the _current_alias."""
         self._current_location = new_location
-        if (self._current_location.query_path, None) in self._aliases:
-            self._current_alias = self._aliases[(self._current_location.query_path, None)]
+        # Create appropriate alias key based on whether or not relocation is taking place in a fold
+        # or not. Note: during end_fold relocation to outside of the fold (i.e. to a Location rather
+        # than a FoldScopeLocation) occurs before self._current_fold is set to None. Therefore, it
+        # should be checked that the current fold is not None AND that the new location is a
+        # FoldScopeLocation.
```
Comment (bojanserafimov) [inline graphql_compiler/compiler/emit_sql.py:718]: Is this function ever called with a `FoldScopeLocation` and  `_current_fold` being `None`? Seems like `fold`, `unfold` and `traverse (inside fold)` all have `_current_fold` set when calling it.

If no, I'd remove `_current_fold` logic from this function.

**Pull request merged**